In [55]:
#!pip install pandas numpy
#!/usr/bin/env python3

import pandas as pd
import numpy as np
import re

In [56]:
df = pd.read_csv(r"C:\Users\GK111\Downloads\audible_uncleaned.csv\audible_uncleaned.csv")
df.head()


name                     author  \
0                  Geronimo Stilton #11 & #12  Writtenby:GeronimoStilton   
1                            The Burning Maze      Writtenby:RickRiordan   
2                                The Deep End       Writtenby:JeffKinney   
3                        Daughter of the Deep      Writtenby:RickRiordan   
4  The Lightning Thief: Percy Jackson, Book 1      Writtenby:RickRiordan   

                    narrator                time releasedate language  \
0      Narratedby:BillLobely   2 hrs and 20 mins    04-08-08  English   
1   Narratedby:RobbieDaymond   13 hrs and 8 mins    01-05-18  English   
2      Narratedby:DanRussell    2 hrs and 3 mins    06-11-20  English   
3  Narratedby:SoneelaNankani  11 hrs and 16 mins    05-10-21  English   
4  Narratedby:JesseBernstein              10 hrs    13-01-10  English   

                           stars   price  
0     5 out of 5 stars34 ratings  468.00  
1   4.5 out of 5 stars41 ratings  820.00  
2   4.5 out of 5 stars38 ratings  410.00  
3   4.5 out of 5 stars12 ratings  615.00  
4  4.5 out of 5 stars181 ratings  820.00

In [57]:
df.sample(10)

name  \
73905                             The Memory of Her Kiss   
55921                         Yes, SIR. (German edition)   
14040                                     What Unites Us   
70864                         Christmas at Promise Lodge   
16642                                       実説 城谷怪談「黒い瞳」   
58724                               AARP: The Other Talk   
28578                                    How Clients Buy   
6769   The Sleeping Beauty and Other Fairy Tales from...   
86344                             Much Ado About Mothing   
20401                          Andy Warhol Was a Hoarder   

                                    author  \
73905               Writtenby:RebeccaRuger   
55921              Writtenby:AngelinaConti   
14040  Writtenby:DanRather,ElliotKirschner   
70864           Writtenby:CharlotteHubbard   
16642                        Writtenby:城谷歩   
58724                  Writtenby:TimProsch   
28578    Writtenby:TomMcMakin,DougFletcher   
6769   Writtenby:ArthurThomasQuiller-Couch   
86344                 Writtenby:JamesLowen   
20401                Writtenby:ClaudiaKalb   

                                 narrator                time releasedate  \
73905             Narratedby:StewartCrank  10 hrs and 33 mins    22-02-22   
55921  Narratedby:KaiSchulz,MayaPoghosyan   8 hrs and 11 mins    06-10-19   
14040                Narratedby:DanRather    7 hrs and 5 mins    10-11-17   
70864               Narratedby:SusanBoyce    9 hrs and 4 mins    27-09-16   
16642                      Narratedby:城谷歩             30 mins    04-08-21   
58724                Narratedby:MarlinMay    5 hrs and 6 mins    28-07-21   
28578              Narratedby:BarryAbrams   6 hrs and 34 mins    28-05-19   
6769                Narratedby:RoeKendall   2 hrs and 50 mins    06-01-05   
86344                Narratedby:MaxDowler  11 hrs and 28 mins    27-05-21   
20401               Narratedby:LisaLarsen  10 hrs and 46 mins    13-07-21   

       language                         stars     price  
73905   English                 Not rated yet    703.00  
55921    german                 Not rated yet    334.00  
14040   English                 Not rated yet    703.00  
70864   English                 Not rated yet  1,172.00  
16642  japanese                 Not rated yet    139.00  
58724   English                 Not rated yet    469.00  
28578   English  4.5 out of 5 stars10 ratings    586.00  
6769    English                 Not rated yet    468.00  
86344   English                 Not rated yet  1,138.00  
20401   English                 Not rated yet    702.00

In [58]:
# Clean author
df['author'] = (
    df['author']
    .str.replace(r'Writtenby:', '', regex=True)   # remove "Writtenby:"
    .str.strip()                                 # remove extra spaces
)

# Clean narrator
df['narrator'] = (
    df['narrator']
    .str.replace(r'Narratedby:', '', regex=True) # remove "Narratedby:"
    .str.strip()
)

print(df['narrator'].sample(10))
print('###################')
print(df['author'].sample(10))

73323                     MelissaMoran
12338     TeriSchnaubelt,GaryTiedemann
68075        Jean-ChristopheFreseuilhe
77823                      NicolePoole
55405                       SimonVance
79259                      CalebScharf
40704    NewtGingrich,CallistaGingrich
41757                      NicoleForup
7920                   CassandraMorris
1510                         anonymous
Name: narrator, dtype: object
###################
28296                  MichelPoulaert
2030     MariSchuh,GailSaunders-Smith
84114        LisaRoecker,LauraRoecker
18582                      JoanDidion
53090                     PedroZuazua
54794                  JulieAnnWalker
45802                TresserHenderson
76123                  JenniferVanWyk
82453                   WilliamHussey
51155                   AllisonPataki
Name: author, dtype: object


In [59]:
#Clean rerease date
df['releasedate'] = pd.to_datetime(
    df['releasedate'],
    errors='coerce',              # invalid dates become NaT (null datetime)
    dayfirst=True                 # important: your sample is day-month-year
)
df['releasedate'].sample(10)

C:\Users\GK111\AppData\Local\Temp\ipykernel_6736\957617465.py:2: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['releasedate'] = pd.to_datetime(


58766   2021-01-05
86722   2005-08-09
3725    2013-01-22
39484   2015-11-18
15204   2015-01-12
71755   2021-08-02
55272   2022-01-25
48997   2021-12-23
1214    2021-05-05
36333   2022-01-13
Name: releasedate, dtype: datetime64[ns]

In [60]:
# Function to convert strings like "2 hrs and 36 mins" to total minutes as int like 156
def convert_time_to_minutes(time_str):
    if pd.isna(time_str):   # Handle missing values
        return None
    
    # Extract hours and minutes using regex
    hours = re.search(r'(\d+)\s*hr', time_str)
    mins = re.search(r'(\d+)\s*min', time_str)
    
    total_minutes = 0
    if hours:
        total_minutes += int(hours.group(1)) * 60   # convert hrs to mins
    if mins:
        total_minutes += int(mins.group(1))         # add mins
    
    return total_minutes


In [61]:
import re

# split stars and ratings properly
# the stars part is like "4.6 out of 5 stars"
# the ratings part is like "1,234 ratings"
# othertimes its "Not rated yet"

def extract_stars(text):
    # if its "Not rated yet" then return nothing
    if "Not rated yet" in text:
        return None
    
    # try to find something like "4.5 out of 5 stars"
    match = re.search(r"(\d+(\.\d+)?) out of 5 stars", text)
    if match:
        return float(match.group(1))  # this gives me the number before "out of 5"
    else:
        return None

def extract_ratings(text):
    # again, ignore "Not rated yet"
    if "Not rated yet" in text:
        return None
    
    # look for number before "ratings"
    # \d means digit, + means one or more, ,? allows commas
    match = re.search(r"([\d,]+) ratings", text)
    if match:
        # remove commas from numbers like "1,234"
        return int(match.group(1).replace(",", ""))
    else:
        return None

# now apply both
df["stars_clean"] = df["stars"].apply(extract_stars)
df["ratings"] = df["stars"].apply(extract_ratings)

# quick check
print(df[["stars", "stars_clean", "ratings"]].head(15))


                            stars  stars_clean  ratings
0      5 out of 5 stars34 ratings          5.0     34.0
1    4.5 out of 5 stars41 ratings          4.5     41.0
2    4.5 out of 5 stars38 ratings          4.5     38.0
3    4.5 out of 5 stars12 ratings          4.5     12.0
4   4.5 out of 5 stars181 ratings          4.5    181.0
5      5 out of 5 stars72 ratings          5.0     72.0
6      5 out of 5 stars11 ratings          5.0     11.0
7      5 out of 5 stars50 ratings          5.0     50.0
8       5 out of 5 stars5 ratings          5.0      5.0
9      5 out of 5 stars58 ratings          5.0     58.0
10  4.5 out of 5 stars130 ratings          4.5    130.0
11      5 out of 5 stars6 ratings          5.0      6.0
12      5 out of 5 stars7 ratings          5.0      7.0
13     5 out of 5 stars41 ratings          5.0     41.0
14      4 out of 5 stars2 ratings          4.0      2.0


In [62]:
# Remove duplicates and reset index
#df.drop_duplicates(inplace=True)
df.reset_index(drop=True, inplace=True)

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 87489 entries, 0 to 87488
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   name         87489 non-null  object        
 1   author       87489 non-null  object        
 2   narrator     87489 non-null  object        
 3   time         87489 non-null  object        
 4   releasedate  87489 non-null  datetime64[ns]
 5   language     87489 non-null  object        
 6   stars        87489 non-null  object        
 7   price        87489 non-null  object        
 8   stars_clean  15072 non-null  float64       
 9   ratings      9250 non-null   float64       
dtypes: datetime64[ns](1), float64(2), object(7)
memory usage: 6.7+ MB


In [63]:
#get a statistical summary of the numerical columns
df.describe(include='all')

name    author   narrator    time  \
count            87489     87489      87489   87489   
unique           82767     48374      29717    2284   
top     The Art of War  矢島雅弘,石橋遊  anonymous  2 mins   
freq                20       874       1034     372   
mean               NaN       NaN        NaN     NaN   
min                NaN       NaN        NaN     NaN   
25%                NaN       NaN        NaN     NaN   
50%                NaN       NaN        NaN     NaN   
75%                NaN       NaN        NaN     NaN   
max                NaN       NaN        NaN     NaN   
std                NaN       NaN        NaN     NaN   

                          releasedate language          stars   price  \
count                           87489    87489          87489   87489   
unique                            NaN       36            665    1011   
top                               NaN  English  Not rated yet  586.00   
freq                              NaN    61884          72417    5533   
mean    2018-06-23 22:53:20.411480064      NaN            NaN     NaN   
min               1998-12-27 00:00:00      NaN            NaN     NaN   
25%               2016-09-06 00:00:00      NaN            NaN     NaN   
50%               2020-02-06 00:00:00      NaN            NaN     NaN   
75%               2021-08-17 00:00:00      NaN            NaN     NaN   
max               2025-11-14 00:00:00      NaN            NaN     NaN   
std                               NaN      NaN            NaN     NaN   

        stars_clean       ratings  
count   15072.00000   9250.000000  
unique          NaN           NaN  
top             NaN           NaN  
freq            NaN           NaN  
mean        4.45694     34.587243  
min         1.00000      2.000000  
25%         4.00000      2.000000  
50%         4.50000      5.000000  
75%         5.00000     14.000000  
max         5.00000  12573.000000  
std         0.72394    264.025444

In [64]:
#check for missing values
df.isnull().sum() 
#stars and ratings seem to have the most missing values

name               0
author             0
narrator           0
time               0
releasedate        0
language           0
stars              0
price              0
stars_clean    72417
ratings        78239
dtype: int64

In [65]:
#check uniqueness as duplicates
for col in df.columns:
    print(col, ":", df[col].nunique(), "unique values")

print('\n duplicated:', df.duplicated().sum())


name : 82767 unique values
author : 48374 unique values
narrator : 29717 unique values
time : 2284 unique values
releasedate : 5058 unique values
language : 36 unique values
stars : 665 unique values
price : 1011 unique values
stars_clean : 9 unique values
ratings : 383 unique values

 duplicated: 0


In [66]:
df.ratings
df.stars_clean

0        5.0
1        4.5
2        4.5
3        4.5
4        4.5
        ... 
87484    NaN
87485    NaN
87486    NaN
87487    NaN
87488    NaN
Name: stars_clean, Length: 87489, dtype: float64

In [67]:
#Do more exploration to see what cleaning is needed
# Check languages - maybe some are written funny (english vs English)
print(df["language"].unique())

# Check prices - are they strings or numeric?
print(df["price"].head())

# Convert price to numeric
df["price"] = pd.to_numeric(df["price"], errors="coerce")

# Sanity check: what's the cheapest and most expensive audiobook?
print("Price min:", df["price"].min())
print("Price max:", df["price"].max())

# Check time column again - make sure all were converted to minutes succcesfully
print(df["time"].sample(10))

# Check duplicates by book name + author (sometimes same title appears multiple times)
dupes = df.duplicated(subset=["name", "author"])
print("Duplicate books:", dupes.sum())


['English' 'Hindi' 'spanish' 'german' 'french' 'catalan' 'swedish'
 'italian' 'danish' 'finnish' 'dutch' 'hebrew' 'russian' 'polish'
 'galician' 'afrikaans' 'icelandic' 'romanian' 'japanese' 'tamil'
 'portuguese' 'urdu' 'hungarian' 'czech' 'bulgarian' 'mandarin_chinese'
 'basque' 'korean' 'arabic' 'greek' 'turkish' 'ukrainian' 'slovene'
 'norwegian' 'telugu' 'lithuanian']
0    468.00
1    820.00
2    410.00
3    615.00
4    820.00
Name: price, dtype: object
Price min: 11.0
Price max: 999.0
27465     5 hrs and 48 mins
51852     4 hrs and 37 mins
56316     6 hrs and 38 mins
12519     9 hrs and 16 mins
6824                 6 mins
40592    42 hrs and 54 mins
16256               17 mins
27810     6 hrs and 12 mins
20112    16 hrs and 16 mins
9689                 2 mins
Name: time, dtype: object
Duplicate books: 1523


Languages - messy casing ('English' vs 'spanish' vs 'german'). Some are lowercase, some are proper, some snake_case (mandarin_chinese).

Price - successfully numeric now , looks sane (11.0 min, 999.0 max).

Time - still a string mix ("59 mins", "7 hrs and 21 mins"). Needs parsing into total minutes.

Duplicates - 1,523 duplicates on name + author. That’s not tragic out of 87k entries, but still dirty.

In [68]:
#  Clean language column (capitalize first letter, replace snake_case)
df["language_clean"] = df["language"].str.strip().str.title() # capitalize first letter of each word
df["language_clean"] = df["language_clean"].str.replace("_", " ") # replace underscores with spaces

print(df["language_clean"].unique()) # check unique values

# Handle duplicates (drop them if exact duplicated)
df = df.drop_duplicates(subset=["name", "author"]) # drop duplicates based on name + author match
print("After dropping duplicates:", df.shape) # check new shape

# Convert time into minutes

def convert_time_to_minutes(x):
    if pd.isna(x): # Handle missing values
        return None # return None if time is missing
    hours = re.search(r"(\d+)\s*hr", x) # look for something like "2 hr" or "1 hr"
    mins = re.search(r"(\d+)\s*min", x) # look for something like "36 min"

    h = int(hours.group(1)) * 60 if hours else 0 # convert hrs to mins
    m = int(mins.group(1)) if mins else 0 # get mins
    return h + m # return total minutes

df["time_mins"] = df["time"].apply(convert_time_to_minutes) # apply function to time column

print("Time in minutes summary:")
print(df["time_mins"].describe()) # check summary stats


['English' 'Hindi' 'Spanish' 'German' 'French' 'Catalan' 'Swedish'
 'Italian' 'Danish' 'Finnish' 'Dutch' 'Hebrew' 'Russian' 'Polish'
 'Galician' 'Afrikaans' 'Icelandic' 'Romanian' 'Japanese' 'Tamil'
 'Portuguese' 'Urdu' 'Hungarian' 'Czech' 'Bulgarian' 'Mandarin Chinese'
 'Basque' 'Korean' 'Arabic' 'Greek' 'Turkish' 'Ukrainian' 'Slovene'
 'Norwegian' 'Telugu' 'Lithuanian']
After dropping duplicates: (85966, 11)
Time in minutes summary:
count    85966.000000
mean       416.873694
std        362.906468
min          1.000000
25%        142.000000
50%        386.500000
75%        584.000000
max       8595.000000
Name: time_mins, dtype: float64


### Do feature Ingineering on the dataset

In [69]:
df.columns

Index(['name', 'author', 'narrator', 'time', 'releasedate', 'language',
       'stars', 'price', 'stars_clean', 'ratings', 'language_clean',
       'time_mins'],
      dtype='object')

In [70]:
# Create column with Book length categories to show short, medium, long and very long books
#create a function to categorize length
def categorize_length(minutes):
    if minutes < 120:   # less than 2 hrs
        return "Short"
    elif minutes < 360: # less than 6 hrs
        return "Medium"
    elif minutes < 720: # less than 12 hrs
        return "Long"
    else:
        return "Very Long"

#call the function to create new column 'length_category'
df["length_category"] = df["time_mins"].apply(categorize_length)

# Release year 
df["release_year"] = pd.to_datetime(df["releasedate"]).dt.year

# create Price buckets to show cheap, mid and premium books
def price_bucket(price):
    if price < 200:
        return "Cheap"
    elif price <= 500:
        return "Mid"
    else:
        return "Premium"

#call the function to create new column 'price_bucket'

df["price_bucket"] = df["price"].astype(float).apply(price_bucket)

# Define Star rating category to show low, medium and high rated books in 3 categories in new column 'star_category'
df["star_category"] = pd.cut(df["stars_clean"],
                             bins=[0, 3, 4, 5],
                             labels=["Low", "Medium", "High"],
                             include_lowest=True)

# create column Rating density based on number of ratings per minute of audiobook length
df["rating_density"] = df["ratings"] / df["time_mins"]

# Get Author productivity based on number of books by author
author_counts = df["author"].value_counts()
df["author_books_count"] = df["author"].map(author_counts)

# Quick check
print(df[["time_mins","length_category","release_year","price","price_bucket",
          "stars_clean","star_category","ratings","rating_density","author","author_books_count"]].head(10))


   time_mins length_category  release_year  price price_bucket  stars_clean  \
0        140          Medium          2008  468.0          Mid          5.0   
1        788       Very Long          2018  820.0      Premium          4.5   
2        123          Medium          2020  410.0          Mid          4.5   
3        676            Long          2021  615.0      Premium          4.5   
4        600            Long          2010  820.0      Premium          4.5   
5        635            Long          2018  656.0      Premium          5.0   
6        143          Medium          2014  233.0          Mid          5.0   
7        752       Very Long          2017  820.0      Premium          5.0   
8        656            Long          2017    NaN      Premium          5.0   
9        802       Very Long          2019  820.0      Premium          5.0   

  star_category  ratings  rating_density           author  author_books_count  
0          High     34.0        0.242857  Geronimo

In [71]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 85966 entries, 0 to 87488
Data columns (total 18 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   name                85966 non-null  object        
 1   author              85966 non-null  object        
 2   narrator            85966 non-null  object        
 3   time                85966 non-null  object        
 4   releasedate         85966 non-null  datetime64[ns]
 5   language            85966 non-null  object        
 6   stars               85966 non-null  object        
 7   price               76673 non-null  float64       
 8   stars_clean         14657 non-null  float64       
 9   ratings             8961 non-null   float64       
 10  language_clean      85966 non-null  object        
 11  time_mins           85966 non-null  int64         
 12  length_category     85966 non-null  object        
 13  release_year        85966 non-null  int32         


In [72]:
# extract year, month, day from the releasedate column
df['year'] = df['releasedate'].dt.year
df['month'] = df['releasedate'].dt.month
df['day'] = df['releasedate'].dt.day

# maybe also get day name, like Monday, Tuesday... for vibes
df['day_name'] = df['releasedate'].dt.day_name()

# let's create weekend flag (1 if Sat/Sun, else 0)
df['is_weekend'] = df['day_name'].isin(['Saturday', 'Sunday']).astype(int)



# create buckets like "short listen" , "medium listen" vs "long listen"
def get_length_category(mins):
    if mins < 60:
        return "very short (<1hr)"
    elif mins < 180:
        return "short (1–3hrs)"
    elif mins < 480:
        return "medium (3–8hrs)"
    else:
        return "long (>8hrs)"

df['listen_length_category'] = df['time_mins'].apply(get_length_category)




In [73]:
#Final sanity check to makesure data is clean and ready for analysis, possible export to csv and modeling
# --- FINAL SANITY CHECKS ---

print("\n✅ Shape of dataset:")
print(df.shape)   # rows, columns

print("\n✅ Columns & dtypes:")
print(df.dtypes)

print("\n✅ Missing values per column:")
print(df.isnull().sum())

print("\n✅ Quick look at numerical columns:")
print(df.describe())

print("\n✅ Quick look at object/categorical columns:")
print(df.describe(include='object'))

print("\n✅ Sample rows (to confirm new features look right):")
print(df[['releasedate','year','month','day','day_name','is_weekend','time_mins','listen_length_category']].head(10))

print("\n✅ Check unique values in new engineered columns:")
print("Day names:", df['day_name'].unique())
print("Weekend flag counts:\n", df['is_weekend'].value_counts())
print("Listen length categories:\n", df['listen_length_category'].value_counts())



✅ Shape of dataset:
(85966, 24)

✅ Columns & dtypes:
name                              object
author                            object
narrator                          object
time                              object
releasedate               datetime64[ns]
language                          object
stars                             object
price                            float64
stars_clean                      float64
ratings                          float64
language_clean                    object
time_mins                          int64
length_category                   object
release_year                       int32
price_bucket                      object
star_category                   category
rating_density                   float64
author_books_count                 int64
year                               int32
month                              int32
day                                int32
day_name                          object
is_weekend                         int64
lis

In [77]:
# Save the cleaned DataFrame to a new CSV file AND SELECTED COLUMNS
# final clean dataset columns
final_cols = [
    "name", "author", "narrator",
    "releasedate", "release_year", "month", "day", "day_name", "is_weekend",
    "language_clean",
    "time_mins", "listen_length_category",
    "price", "price_bucket",
    "stars_clean", "star_category",
    "ratings", "rating_density",
    "author_books_count"
]

# subset df
df_clean = df[final_cols].copy()

# sanity check shape
print("Final clean shape:", df_clean.shape)

# save to CSV
df_clean.to_csv("books_clean.csv", index=False)



Final clean shape: (85966, 19)


#### Boom Shakalaka!!! ADIOS!!!!!!!!